opentelemetry-sdk/src/export/trace.rs

```diff
@@ -10,6 +10,9 @@ use std::time::SystemTime;
 /// Describes the result of an export.
 pub type ExportResult = Result<(), TraceError>;
 
+/// Describes the results of other operations on the trace API.
+pub type TraceResult<T> = Result<T, TraceError>;
```

Comment:
Need to ensure consistency here. LogExporter uses same LogResult for export and shutdown both, but TraceExporter uses different Results.

Is this something we can fix too?

Reply:
We can use a type alias so that `export` is marked as returning `TraceResult<()>`, but the existing callers (all 38 of them in our codebase) don't have to change. 

I've pushed a change that works this way. I can also just pull of the band-aid and fix all our existing internal usages to use `TraceResult` directly. 

As we are looking at consistency here, another thing that jumps out is that we're using `TraceError` (which i've introduced `TraceResult` to match) , despite the trait itself being `SpanExporter`. Better would be `SpanError` and `SpanResult`. To give an indication of scale, we have 72 refs in this codebase to `TraceError` alone (plus external usages of course).

My suggestion is:

* Leave `TraceError` alone
* Change our internal usages of `ExportResult` to `TraceResult<()>` 
* Leave the `ExportResult` alias for any external usages but mark it as deprecated

Thoughts @cijothomas ?

Reply:
I'll need some time to think through this. I am unsure if the current state is what we finally want. I am thinking that we should have very specific Enums for each Result returned to user.
If shutdown() is called on an already shutdown provider, then the Result should an Error variant, but a specific one like Error::ShutdownAlreadyInvoked, so users know exactly what failed and why instead of relying on "Other("something went wrong")"

This will require more refactoring, but something we should be going before Stable. 
The "Other" should be truly for "others" - i.e something went wrong that is not accounted for, and should be rarely used, if not possible to avoid entirely.

Reply:
Another quick thought - validate if it is considered breaking to expand the enum with new variants. Do we need to provide Into/From for std Error and that is sufficient?

Reply:
Ok - so we're happy to take some breaking changes now and make things clean! 

I think what you'd typically see is errorType-per-function, with explicitly mapped errors where it makes sense (e.g. youe example with shutdown-already-invoked) and  [thiserror](https://docs.rs/thiserror/latest/thiserror/)  used to automatically map the rest (things like socket errors, etc. - where re-wrapping them makes no sense).

These days you see `anyhow` used a lot for application code (e.g., see how i've added it to the integration tests to clean up error handling) and `thiserror` used a lot in library code, or in situations where you care about being able to map on top of the error variants.

I think this would be pretty idiomatic (note - errors and logic for illustration purposes  only):

```rust

use thiserror::Error;

#[derive(Error, Debug)]
pub enum ExportError {
     #[error("Socket IO Error")]
    SocketError(#[from] io::Error),

    #[error("Exporter already shut down")]
    ExporterShutdown,
}

pub trait SpanExporter {
    async fn export(&mut self, batch: Vec<SpanData>) -> Result<(), ExportError> { 
          // explicit error
          if already_shutdown { 
              Err(ExportError::ExporterShutdown);
          }

          // by way of example, imagine this may return an IO error
          // thiserror has setup the error mapping for us, so we can just use ? 
          let ack = my_socket.send()?;
         
          Ok(())
     }
    async fn shutdown(&mut self) -> Result<(), ShutdownError>; 
}
```

Finally I don't think the extra typedefs add much value.

Reply:
Not sure if I understood this fully, but the code already uses `thiserror` nicely to get the error status from exporters, with below skeleton. Are you proposing to change this in better way?

```rust
#[derive(thiserror::Error, Debug)]
#[non_exhaustive]
pub enum TraceError {
    #[error("Exporter {0} encountered the following error(s): {name}", name = .0.exporter_name())]
    ExportFailed(Box<dyn ExportError>),
    // Rest of variants
}
/// Trait for errors returned by exporters
pub trait ExportError: std::error::Error + Send + Sync + 'static {
    /// The name of exporter that returned this error
    fn exporter_name(&self) -> &'static str;
}
// and then in OTLP  crate
/// Wrap type for errors from this crate.
#[derive(thiserror::Error, Debug)]
pub enum Error {
    // OTLP specific error variants
}
impl opentelemetry::trace::ExportError for Error {
    fn exporter_name(&self) -> &'static str {
        "otlp"
    }
}
``

Reply:
Whoops, so it does. i'll fix my suggestion so it doesn't include adding thiserror as its already there.

Reply:
// explicit error if already_shutdown { Err(ExportError::ExporterShutdown) }

^ I actually meant to suggest we use explicit ones like this, instead of "Other("Shutdown. already invoked"), consistently throughout.

Reply:
@cijothomas @lalitb I think this makes sense, otherwise we're assuming a lot about what custom implementations might need in terms of thiserror-mapped-errors.

Are we happy for me to go through and apply this? Would look like:

 ```rust
shutdown(...) -> Result<(), ShutdownError>
export(...) -> Result<(), ExportError>
```

and an extension of the error types following the explicit-named-errors-style established already in `LogError`, e.g.:

```rust
pub enum LogError {
    /// Export failed with the error returned by the exporter.
    #[error("Exporter {0} encountered the following errors: {name}", name = .0.exporter_name())]
    ExportFailed(Box<dyn ExportError>),

    /// Export failed to finish after certain period and processor stopped the export.
    #[error("Exporter timed out after {} seconds", .0.as_secs())]
    ExportTimedOut(Duration),

    /// Processor is already shutdown
    #[error("{0} already shutdown")]
    AlreadyShutdown(String),

    /// Mutex lock poisoning
    #[error("mutex lock poisioning for {0}")]
    MutexPoisoned(String),
```